### PR TITLE
[@types/bootstrap] add definition for getOrCreateInstance

### PR DESCRIPTION
**Pull request by darthf1**
Resolves:
https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/54828
https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/54289
https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/54026

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://getbootstrap.com/docs/5.1/components/alerts/#methods
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.

### CODE REVIEW

**Review by martin-badin — APPROVED** [review]

**Review by martin-badin — CHANGES_REQUESTED** [review]
Please add tests

**Comment by darthf1** [comment]
@martin-badin thanks for the review. I fixed the lint errors and added tests. I added a `GetOrCreateInstanceFactory` type because the function should only return a component instance (it is created if not exists) and not null.

File: types/bootstrap/js/dist/base-component.d.ts [hunk 1]
```diff
@@ -3,6 +3,7 @@
  * each plugin exposes a getInstance method. In order to retrieve it directly from an element, do this: bootstrap.Popover.getInstance(myPopoverEl)
  */
 export type GetInstanceFactory<T> = (element: string | Element) => T | null;
+export type GetOrCreateInstanceFactory<T> = (element: string | Element) => T;
```
Comment (martin-badin) [inline types/bootstrap/js/dist/base-component.d.ts:6]: The method has two patameters. Please see bootstrap repository and add the second parameter. 

static getOrCreateInstance(element, config = {}) {
    return this.getInstance(element) || new this(element, typeof config === 'object' ? config : null)
  }

**Review by martin-badin — CHANGES_REQUESTED** [review]

File: types/bootstrap/js/dist/base-component.d.ts (hunk 1 above)
Comment (darthf1) [reply types/bootstrap/js/dist/base-component.d.ts]: see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/55213/commits/5e19743e2b6586383a74b26ee6debcfe1c7a12e2

File: types/bootstrap/js/dist/alert.d.ts [hunk 2]
```diff
@@ -4,6 +4,7 @@ declare class Alert extends BaseComponent {
     static NAME: 'alert';
     static jQueryInterface: Alert.jQueryInterface;
     static getInstance: GetInstanceFactory<Alert>;
+    static getOrCreateInstance: GetInstanceFactory<Alert>;
```
[the remark this answers is not in the document]
Comment (darthf1) [reply types/bootstrap/js/dist/alert.d.ts]: see https://github.com/DefinitelyTyped/DefinitelyTyped/pull/55213/commits/5e19743e2b6586383a74b26ee6debcfe1c7a12e2

**Comment by darthf1** [comment]
I added the second parameter as
```ts
export type GetOrCreateInstanceFactory<T> = (element: string | Element, config?: Object) => T;
```

Or would you prefer
```ts
export type GetOrCreateInstanceFactory<T> = (element: string | Element, config?: { [key: string]: any }) => T;
```

**Comment by tommie** [comment]
There's already (at least) a `Modal.Options` type. I'm guessing that should be used for `Modal`.

**Comment by martin-badin** [comment]
> I added the second parameter as
> 
> ```ts
> 
> export type GetOrCreateInstanceFactory<T> = (element: string | Element, config?: Object) => T;
> 
> ```
> 
> 
> 
> Or would you prefer
> 
> ```ts
> 
> export type GetOrCreateInstanceFactory<T> = (element: string | Element, config?: { [key: string]: any }) => T;
> 
> ```

I prefer Record<string, any>

**Comment by darthf1** [comment]
@martin-badin all done!

**Review by martin-badin — APPROVED** [review]
Looks good

**Comment by darthf1** [comment]
Ready to merge

**Pull request merged**